packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,67 +165,213 @@ function makeSwingStore(dirPath, forceReset = false) {
     delete: del,
   };
 
+  /** @type {Set<number>} */
+  const activeStreamFds = new Set();
+  /** @type {Map<string, number>} */
+  const streamFds = new Map();
+  /** @type {Map<string, string>} */
+  const streamStatus = new Map();
+  let statusCounter = 0;
+
+  const STREAM_START = harden({ offset: 0, itemCount: 0 });
+
+  function insistStreamName(streamName) {
+    assert.typeof(streamName, 'string');
+    assert(
+      streamName.match(/^[-\w]+$/),
+      X`invalid stream name ${q(streamName)}`,
+    );
+  }
+
+  function closefd(fd) {
+    try {
+      fs.closeSync(fd);
+    } catch (e) {
+      // closing an already closed fd is OK, but any other errors are probably bad
+      if (e.code !== 'EBADF') {
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Close a stream that's open for read or write.
+   *
+   * @param {string} streamName  The stream to close
+   */
+  function closeStream(streamName) {
+    insistStreamName(streamName);
+    const fd = streamFds.get(streamName);
+    if (fd) {
+      closefd(fd);
+      streamFds.delete(streamName);
+      activeStreamFds.delete(fd);
+      streamStatus.set(streamName, 'unused');
+    }
+  }
+
   /**
    * Generator function that returns an iterator over the items in a stream.
    *
    * @param {string} streamName  The stream to read
+   * @param {Object} endPosition  The position of the end of the stream
+   * @param {Object?} startPosition  Optional position to start reading from
    *
-   * @yields {string} an iterator for the items in the named stream
+   * @returns {Iterable<string>} an iterator for the items in the named stream
    */
-  function* readStream(streamName) {
-    try {
-      const fd = fs.openSync(`${dirPath}/${streamName}`, 'r');
-      if (fd) {
-        const reader = new Readlines(fd);
-        while (true) {
-          const line = /** @type {string|false} */ (reader.next());
-          if (line) {
-            yield line;
-          } else {
-            break;
+  function openReadStream(streamName, endPosition, startPosition) {
```

Comment:
This might be simpler/more-reliable if we make both `startPosition` and `endPosition` mandatory (and if so, swap them, of course). The VatWarehouse can set both to `streamStore.STREAM_START` when the vat is created. It will increment `endPosition` as transcript entries are written, and increment `startPosition` when heap snapshots are made. Tidy and fewer cases to reason about.

Reply:
I was trying to make it simple for the usual case of reading from the beginning, but I think you're right in that there's not really much percentage in defaulting common-case parameters when the API is only going to be used from a very small number of places by code we entirely control.  (Actually, now that I think about it, there are going to be a bunch of calls to `openReadStream`, almost all of which are going to be in test code.  It's probably not a sound idea to optimize API design for the convenience of writing tests to test it.)

Reply:
Hah. Yep, I'm guilty of optimizing for tests.. had to rewrite a lot of APIs once they hit real usage because of that one :).